cleo_plugins/Audio/CSoundSystem.cpp

```diff
@@ -0,0 +1,459 @@
+    }
+
+    void CAudioStream::SetSpeed(float val, float transitionTime)
+    {
+        if (val < 0.0f) // just update
+        {
+            if (speed != speedTarget)
+            {
+                auto timeDelta = CTimer::m_snTimeInMillisecondsNonClipped - CTimer::m_snPreviousTimeInMillisecondsNonClipped;
+                speed += speedTransitionStep * (double)timeDelta; // animate the transition
+
+                // check progress
+                auto remaining = speedTarget - speed;
+                remaining *= (speedTransitionStep > 0.0) ? 1.0 : -1.0;
+                if (remaining < 0.0) // overshoot
+                {
+                    speed = speedTarget; // done
+                    if (speed <= 0.0f) Pause();
+                }
+            }
+        }
+        else
+        {
+            if (val > 0.0f) Resume();
+
+            speedTarget = val;
+            if (transitionTime <= 0.0) speed = val; // instant
+            else speedTransitionStep = (speedTarget - speed) / (1000.0 * transitionTime);
+        }
+
+        float freq = rate * (float)speed * CSoundSystem::speed;
+        freq = max(freq, 0.000001f); // 0 results in original speed
+        BASS_ChannelSetAttribute(streamInternal, BASS_ATTRIB_FREQ, freq);
+    }
+
+    float CAudioStream::GetSpeed() const
+    {
+        return (float)speed;
+    }
+
+    void CAudioStream::Loop(bool enable)
+    {
+        BASS_ChannelFlags(streamInternal, enable ? BASS_SAMPLE_LOOP : 0, BASS_SAMPLE_LOOP);
+    }
+
+	HSTREAM CAudioStream::GetInternal()
+	{
+		return streamInternal;
+	}
+
+    void CAudioStream::Process(bool force)
+    {
+        switch (BASS_ChannelIsActive(streamInternal))
+        {
+            case BASS_ACTIVE_PAUSED:
+                state = paused;
+                break;
+
+            case BASS_ACTIVE_PLAYING:
+            case BASS_ACTIVE_STALLED:
+                state = playing;
+                break;
+
+            case BASS_ACTIVE_STOPPED:
+                state = stopped;
+                break;
+        }
+
+        if (state != playing && !force) return; // done
+
+        SetSpeed(-1.0f);
+        SetVolume(-1.0f);
+    }
+
+    void CAudioStream::Set3dPosition(const CVector& pos)
+    {
+        // not applicable for 2d audio
+    }
+
+    void CAudioStream::Set3dSize(float radius)
+    {
+        // not applicable for 2d audio
+    }
+
+    void CAudioStream::Link(CPlaceable *placable)
+    {
+        // not applicable for 2d audio
+    }
+
+    void C3DAudioStream::Set3dPosition(const CVector& pos)
+    {
+        link = nullptr;
+
+        position.x = pos.y;
+        position.y = pos.z;
+        position.z = pos.x;
+        BASS_3DVECTOR vel = { 0.0f, 0.0f, 0.0f };
+        BASS_ChannelSet3DPosition(streamInternal, &position, nullptr, &vel);
+    }
+
+    void C3DAudioStream::Set3dSize(float radius)
+    {
+        BASS_ChannelSet3DAttributes(streamInternal, BASS_3DMODE_NORMAL, radius, 1E+12f, -1, -1, -1.0f);
```

Comment:
`1E+12f` ?? 5000.0 covers the entire map

Reply:
BASS internally uses some weird decay curve calculations. I don't know what they used for this parameter as default but it seems to be insufficient. This parameter only stops volume decay after that distance is reached.